numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1846,6 +1846,16 @@ array_empty_like(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kwds)
                                             order, dtype, subok);
     Py_DECREF(prototype);
 
+    /* gh-7854: explicitly fill in Py_None for object dtype,
+       otherwise we get weird data corruption
+    */
```

Comment:
Fair point.  I moved it into `PyArray_NewLikeArray` for the time being, and tests seem to pass locally, though we'll see what Travis and Appveyor have to say about that.  Not sure if I can move it into say `PyArray_NewFromDescr`, as that touches a lot more places than the former, though I can certainly try. 

Confused what you mean in your second sentence.  Are you referring to `np.empty()`?  If I leave the filling at the `PyArray_NewLikeArray` level, `np.empty()` will not see this and can be left unchanged.  However, if I do move it to the `PyArray_NewFromDescr` level, then I will have to add some more conditionals to make sure that we aren't filling in NULL twice.